azafea/event_processors/endless/metrics/machine.py

```diff
@@ -67,3 +73,13 @@ def upsert_machine_live(dbsession: DbSession, machine_id: str) -> None:
                                       set_={'live': True})
 
     dbsession.connection().execute(stmt)
+
+
+def upsert_machine_location(dbsession: DbSession, machine_id: str,
+                            info: Union[Dict[str, Any], List[Any]]) -> None:
+    if isinstance(info, dict):
```

Comment:
Great, looks like inverting the sense of this check did the right thing! Fascinating that mypy can't apply the runtime type check when it's an early return.

I suggest that a comment about this more-general type plus runtime check would be good. Perhaps something like (line wrapped):

```suggestion
    """
    Update the relevant Machine record with information from a LocationLabel event.
    
    Although the info is guaranteed to be a `Dict[str, str]` when received from the client, the database column is of type JSONB, so the `LocationLabel.info` field has a more general type. As a result, the function parameter also has a more general type, and this runtime type check is needed to satisfy mypy.
    """
    if isinstance(info, dict):
```

Reply:
> Great, looks like inverting the sense of this check did the right thing! Fascinating that mypy can't apply the runtime type check when it's an early return.

After reading the previous error again, the problem was not because mypy didn’t understand the `not isinstance` part, but because I had left a comment to ignore types even if they were OK (yes, mypy checks that comments are actually useful oO). I’ve removed the `not` because it was better this way.

> I suggest that a comment about this more-general type plus runtime check would be good. Perhaps something like (line wrapped):

You’re right, that’s a nice gift for the future person who will read this.